packages/flutter/lib/src/material/button_style_button.dart

```diff
@@ -191,33 +194,49 @@ abstract class ButtonStyleButton extends StatefulWidget {
 ///  * [TextButton], a simple button without a shadow.
 ///  * [ElevatedButton], a filled button whose material elevates when pressed.
 ///  * [OutlinedButton], similar to [TextButton], but with an outline.
-class _ButtonStyleState extends State<ButtonStyleButton> with MaterialStateMixin, TickerProviderStateMixin {
-  AnimationController? _controller;
-  double? _elevation;
-  Color? _backgroundColor;
+class _ButtonStyleState extends State<ButtonStyleButton> with TickerProviderStateMixin {
+  AnimationController? controller;
+  double? elevation;
+  Color? backgroundColor;
+  late final MaterialStatesController statesController;
+
+  void handleStatesControllerChange() {
+    // Force a rebuild to resolve MaterialStateProperty properties
+    setState(() { });
+  }
+
+  void initStatesController() {
+    statesController = widget.statesController ?? MaterialStatesController();
+    statesController.update(MaterialState.disabled, !widget.enabled);
+    statesController.addListener(handleStatesControllerChange);
+  }
 
   @override
   void initState() {
     super.initState();
-    setMaterialState(MaterialState.disabled, !widget.enabled);
+    initStatesController();
   }
 
   @override
   void dispose() {
-    _controller?.dispose();
+    statesController.removeListener(handleStatesControllerChange);
+    controller?.dispose();
```

Comment:
I don't think you want to call `dispose` on this unless you allocated it here. Isn't the caller responsible for calling `dispose` if they supply one?

There's a pattern I use for this in several places:

```dart
class FooManager {
  @mustCallSuper
  void dispose() {}
}

class Foo extends StatefulWidget {
  const Foo({this.manager});

  final FooManager? manager;

  @override
  State<Foo> createState() => _FooState();
}

class _FooState extends State<Foo> {
  FooManager? _internalManager;
  FooManager get manager => widget.manager ?? _internalManager!;

  @override
  void initState() {
    super.initState();
    if (widget.manager == null) {
      _internalManager = FooManager();
    }
  }

  @override
  void dispose() {
    _internalManager?.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(Foo oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.manager != oldWidget.manager) {
      if (widget.manager != null) {
        _internalManager?.dispose();
        _internalManager = null;
      } else {
        _internalManager ??= FooManager();
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return const SizedBox();
  }
}
```

And then just call the `manager` getter in the rest of the `State` code.

Reply:
`controller?.dispose();` disposes the button's AnimationController, not its MaterialStatesController. I wasn't disposing the statesController, but I suppose I should.